File: app/src/main/java/com/stm/evenote/rsync/model/SyncFile.java
package com.stm.evenote.rsync.model;

import java.io.File;
import java.util.Objects;
import java.util.function.Supplier;

import org.apache.commons.codec.binary.Hex;

public class SyncFile {

  private SyncPath filename;

  /**
   * MD5 hash alg. is used by evernote as well
   * So we can use it for quick change detection
   * to save bandwidth.
   */
  private String md5Hash;

  private Supplier<byte[]> dataSupplier;

  private long timestamp;

  public boolean matchesHash(SyncFile other) {

    return Objects.equals(this.md5Hash, other.md5Hash);
  }

  public boolean matchesPath(SyncFile other) {

    return Objects.equals(this.filename, other.filename);
  }

  public long getTimestamp() {
    return timestamp;
  }

  public SyncPath getFilename() {

    return filename;
  }

  public SyncFile withSyncPath(SyncPath filename) {

    this.filename = filename;
    return this;
  }

  public SyncFile withPath(String path) {

    this.filename = SyncPath.EMPTY.append(path);
    return this;
  }

  public Supplier<byte[]> getDataSupplier() {

    return dataSupplier;
  }

  public SyncFile withDataSupplier(Supplier<byte[]> dataSupplier) {

    this.dataSupplier = dataSupplier;
    return this;
  }

  public String getMd5Hash() {

    return md5Hash;
  }

  public SyncFile withHash(String hash) {

    this.md5Hash = hash;
    return this;
  }

  public SyncFile withHash(byte[] hash) {

    this.md5Hash = Hex.encodeHexString(hash);
    return this;
  }

  public SyncFile withTimestamp(long timestamp) {
    this.timestamp = timestamp;
    return this;
  }

  public File toFile(String workingDirectory) {

    return new File(this.filename.prepend(workingDirectory).toString());
  }

  @Override
  public String toString() {

    return this.filename + " (MD5:" + this.md5Hash + ")";
  }
}

File: app/src/main/kotlin/com/stm/evenote/rsync/model/SyncFile.kt
package com.stm.evenote.rsync.model

import org.apache.commons.codec.binary.Hex
import java.util.*

class SyncFile {

    var filename: SyncPath = SyncPath.EMPTY
    var timestamp: Long = 0
    var md5Hash: String = ""
    var dataSupplier: () -> ByteArray = { ByteArray(0) }

    fun matchesHash(other: SyncFile): Boolean {
        return Objects.equals(this.md5Hash, other.md5Hash)
    }

    fun equalsPath(otherFile: SyncFile): Boolean {
        return Objects.equals(this.filename, otherFile.filename)
    }

    fun matches(pattern: String): Boolean {
        return this.filename.matches(pattern)
    }

    fun withSyncPath(filename: SyncPath): SyncFile {
        this.filename = filename
        return this
    }

    fun withPath(path: String): SyncFile {

        this.filename = SyncPath.EMPTY.append(path)
        return this
    }

    fun withDataSupplier(dataSupplier: () -> ByteArray): SyncFile {
        this.dataSupplier = dataSupplier
        return this
    }

    fun withHash(hash: String): SyncFile {

        this.md5Hash = hash
        return this
    }

    fun withTimestamp(timestamp: Long): SyncFile {
        this.timestamp = timestamp
        return this
    }

    fun withHash(hash: ByteArray): SyncFile {

        this.md5Hash = Hex.encodeHexString(hash)
        return this
    }

    fun getData(): ByteArray {
        return dataSupplier.invoke()
    }

    override fun toString(): String {
        return this.filename.toString() + " (MD5:" + this.md5Hash + ")"
    }
}